### ISSUE TITLE
Consider making `--arch ARCH` equivalent to `-r CURRENTCFGPAIR-ARCH`

### ISSUE THREAD

**Issue by jnpkrn** [rpm-software-management/mock#435]
### Short description of the problem

I wanted to build an `i686` package on `x86_64` native machine.

Per skimming `mock(1)`, it looked as if `--arch i686` what I need.

Alas, in Fedora context and with this change implemented:

https://fedoraproject.org/wiki/Changes/Noi686Repositories

it apparently cannot work:

```
Errors during downloading metadata for repository 'fedora':
  - Status code: 404 for https://mirrors.fedoraproject.org/metalink?repo=rawhide&arch=i386
```

Provided that the `--arch ARCH` form is not accompanied with an explicit
`-r CFG` specification, are the two variants of execution as in the
title really meant to be semantically _that_ different, or can `--arch`
form be freely translated into the other form in that case?

It would definitely help for this very use case, I believe, since:

```
$ grep url /etc/mock/fedora-rawhide-i386.cfg 
baseurl=https://kojipkgs.fedoraproject.org/repos/rawhide/latest/i386
baseurl=https://kojipkgs.fedoraproject.org/repos/rawhide/latest/src/
```

Thanks for considering.

**Comment by xsuchy** [comment]
`-r fedora-rawhide-i386` set the personality - in fact, the personality is driven by `config_opts['target_arch'] = 'i686'`. There is no need to call it explicitly. The command-line options is there for the case you want to override it e.g., set it to i586, or for bootstrapping completly new architecture.

`--arch ARCH` cannot be equivalent to just changing arch in name of config as the config do more than that e.g., see the differences in baseurl in fedora-31-i386.cfg vs. fedora-31-x86_64.cfg

**Comment by jnpkrn** [comment]
Yes, `baseurl` is all what's this is about, and that `$arch` variable
in there collides with that pattern drop for `i686` on Fedora release
side I've mentioned.

Point is that with that change, `--arch ARCH` became rather pointless
for non-isolated cases (which is a major use case, I presume), at least
I was unable to get it to work even if this is expected to suffice,
deduced from explanations in `mock(1)`:

```
mock --rebuild PKG.src.rpm --arch i386 -a https://kojipkgs.fedoraproject.org/repos/rawhide/latest/i386 --disablerepo=fedora
```

error then is:

```
Error: There are no enabled repositories in "/var/lib/mock/fedora-rawhide-i386/root/dev/null".
```

which doesn't make any sense, since an alternative repo has
just been provided, and I presume it is natural for it to be enabled
(the label by which it is added is unknown, anyway).

Perhaps it make sense to at least capture a coincidence of these:

- alternative architecture detected via `--arch ARCH`,
  where `ARCH` is different from the actual host arch

- problems in builddeps downloads detected

and respond with:

```
Problems in fetching build prerequisites detected, but this attempt to
build in a non-native architecture environment of ARCH may, for some
use cases, as well be restated as a following command that is supposed
to use the appropriate repository URL preset:

mock -r CURRENTCFGPAIR-REQUESTEDARCH ORIGINALARGS
```

But this might be a very private convention of mock configs for Fedora
that it is formed as a triple like this.

I dunno, just pointing out the usability problem I've run into
and others might too.

To justify my initial reasoning some more:
I can imagine running `mock` without `-r` is a very convenient thing to
do because you can leave complexity of entering a correct config formula
behind, and having `--arch` to act as a reasonable preselected config
modifier under the hood would keep that complexity behind, and I suppose
it even originally acted _almost_ like this, until the Fedora RCM change
broke the use case as mentioned.

**Comment by xsuchy** [comment]
> Yes, baseurl is all what's this is about.

Nope. That is just the case of rawhide. But check the 31 config. They differ a lot.

`-a` has effect only when using `--chain` command (for historical reasons). That should be mentioned in the man page. Sorry. Fixed in d28bc4a

> I can imagine running mock without -r is a very convenient thing to
> do because you can leave complexity of entering a correct config formula

I will argue the other way round. You should not need to use `--arch` at all. I never used that in my whole life. You want to use the shipped config. Or alter those config. But once you leave out the config, you are in swaps where every step can take you down.

**Comment by jnpkrn** [comment]
On 06/12/19 03:46 -0800, Miroslav Suchý wrote:
>> Yes, baseurl is all what's this is about.
> 
> Nope. That is just the case of rawhide. But check the 31 config.
> They differ a lot.
> 
> `-a` has effect only when using `--chain` command (for historical
> reasons). That should be mentioned in the man page. Sorry. Fixed in
> d28bc4a

Oh, thanks then :-)

Hopefully it will redirect anyone with the same intentions right
to the config route -- I believe it'd do the trick for me.

-- 
Jan (Poki)
